### ISSUE TITLE
Considerations on user-account puppeting

### ISSUE THREAD

**Issue by leonerd** [matrix-org/matrix-appservice-gitter#31]
Users have expressed an interest in being able to give the bridge their gitter-side user credentials so that it can pupper for their actual user account instead of relaying messages via `matrixbot`.

**Comment by leonerd** [comment]
This looks superficially simple but comes with huge amounts of hidden complexity.

Currently the bridge has an easy life, because any messages appearing on either side that come in from humans (or at least, entities external to the bridge) get relayed onto the other side by entities entirely under its control - the `matrixbot` relaybot user on Gitter, or the various `@gitter_*` ghosts on Matrix. This means that any given user-account only requires one-way sync. Whatever a real Matrix or Gitter-side "human" does, the bridge just does the same thing with one of its ghosts on the other side and all is good.

If we were to allow user-account puppeting, there are lots of new situations that arise that require consideration, because there are humans operating both sides of the same "linked user entity" and that may put the bridge into a situation it cannot easily synchronise. A non-exhaustive list of some of them that come to mind are:

1. When a real gitter user uses their real gitter client to speak in a bridged room - how do we relay that back to matrix?
1. When a real gitter user leaves a gitter room in their real gitter client - if we're performing full user-list syncing surely we'd just force them back in again?
1. When a real gitter user for which we are puppeting joins a new gitter-side room that we happen to be bridging into matrix already - what should the matrix side look like?
1. At the moment the user sets up this link between gitter and matrix accounts, what happens wrt. userlist syncing? Does their gitter-side user now join every gitter-side room that the bridge has linked just because their matrix-side user is in the other side of the room?
1. Related to above, what happens at the time the user *disassociates* their accounts - would the bridge now have to puppet the gitter-side user to leave any gitter-side rooms that were only joined because of that previous association? What if the user wanted to remain there?
1. Lets look at the mirror situation - why is the matrix-side account the "primary" and the gitter-side one a relay? What if the user wants the gitter-bridge to puppet their *matrix* account to relay things that they are doing in the gitter side, generally. Joining rooms and so on. If the user is happy to allow the bridge to puppet their gitter account, why not puppet their matrix account too?

These are questions that are new to the Gitter bridge for which there is no precedent in the IRC bridge (because the bridge would never be able to puppet the same IRC-side user account while the user is using it, as IRC does not allow multiple connections of the same user), nor in the Slack or other bridges (because we haven't implemented similar functionality there).

**Comment by yangm97** [comment]
I would say users, at least I, would expect matrix to act as a gitter client. I know, that's not how bridges work, but I think this would cause the least trouble.

**Comment by PureTryOut** [comment]
Definitely. This would also be used in the case of [Telegram](https://github.com/sijmenschoon/telematrix), to make 1-on-1 conversations to Telegram users possible from Matrix.

**Comment by leonerd** [comment]
I've moved out the question of puppeting the matrix-side accounts into a new issue - #33.

**Comment by leonerd** [comment]
Right. That list of questions above has now been edited quite a bit, some of the questions moved out to #33. All of the above remain valid and still require resolving before I can make much progress.

**Comment by leonerd** [comment]
Speaking of progress however, I do have the very start of some experimenal code, on https://github.com/matrix-org/matrix-appservice-gitter/tree/paul/users-and-ghosts

Right now it requires a bunch of hand-crafted `curl` requests to perform the OAuth2 dance to get a user token and then hand-editing of `vim user-store.db` to put that token in the database, but if one is found there the bridge will now use it when relaying messages from matrix-side users. This code does not yet address any of the concerns listed above.

Nor does it begin to address how regular users will actually complete that OAuth2 dance for themselves. This too is a question that needs answering.

**Comment by Kegsay** [comment]
> At the moment the user sets up this link between gitter and matrix accounts, what happens about userlist syncing? Does their gitter-side user now join every gitter-side room that the bridge has linked just because their matrix-side user is in the other side of the room? Or do we join them lazily the first time the matrix-side user speaks?

Upon a successful link to a "real" gitter account, I would treat that real account as the ghost, replacing it in all rooms automatically. So yes, their gitter-side user now joins every bridged room. This has perf implications due to the churn of leaves/joins this would create though, so a lazy-option may be nice.

The fact remains though that the human has explicitly said to the bridge "use and abuse my real gitter user please", so we should honour that and not use the ghost anymore IMO.

> Related to above, what happens at the time the user disassociates their accounts - would the bridge now have to puppet the gitter-side user to leave any gitter-side rooms that were only joined because of that previous association? What if the user wanted to remain there?

Replace the real user with the ghosted user as if they never associated their account in the first place.

> Existing user accounts are likely to already have a "displayname" set in both systems. At the moment of linking how do we cope if they are not identical? If one changes going forward, do we attempt to synchronise it with the other?

Have it as an option to sync across the display name (prefer: Gitter / prefer: Matrix / prefer: Neither (no sync)).

> Harder than a displayname is the avatar image. With a displayname we can just compare strings to see if they're equal, but with avatar images there might be serverside compression or scaling going on that means the "same" image is no longer bytewise identical.

Cheat. There are 3 options:
 - Copy the Gitter image to Matrix.
 - Copy the Matrix image to Gitter.
 - Do not copy.

These don't require comparisons to perform. You just need to fetch the bytes, send them over, then mark that you have synced them.

**Comment by leonerd** [comment]
An initial attempt at this is now running in production, from the https://github.com/matrix-org/matrix-appservice-gitter/tree/paul/users-and-ghosts branch.

This does not attempt to address any concerns other than the minimal required to echo individual message lines into Gitter. In particular:

 * It will not cause the puppeted gitter user account to join or leave any room, for any reason
 * It will not attempt to synchronise displayname or avatar image changes

In addition, there's currently no place on the Riot web UI for users to begin the OAuth2 authorization flow, so to obtain the starting URL for that people will have to ask me to generate one and I'll send it via a 1:1 room.

**Comment by leonerd** [comment]
Latest excitement: if the bridge isn't going to puppet room joins, then it can get stuck in a situation where a puppeted user needs to talk in a gitter room their user account isn't in. I suspect the only sensible thing to do here is for it to revert to using the relaybot as would be the case if it didn't have that user-specific puppet.

This leads to the situation where we need to keep watch for the puppeted users joining/leaving rooms, so as to keep an accurate idea of what rooms we *can* puppet user accounts into

**Comment by leonerd** [comment]
Well. Hrm. Except it turns out that public gitter rooms allow you to post into them regardless of whether you're currently a member of it anyway. As if an IRC channel set to `-n`. So we don't need to check after all.

**Comment by jcgruenhage** [comment]
Any progress on that?

**Comment by devnoname120** [comment]
@leonerd What is the status on this issue? Gitter users regularly complain that the integration is suboptimal.

![image](https://user-images.githubusercontent.com/2824100/34735271-2c311bca-f56f-11e7-8ff6-753ddba4e05e.png)
![image](https://user-images.githubusercontent.com/2824100/34735233-0cbe4718-f56f-11e7-94a1-bf9f166a3f55.png)

**Comment by leonerd** [comment]
I'm afraid I haven't been working for Matrix for the past few months any more.

Last time I was working on it, the bridge itself has all the right pieces, and it was just waiting for someone who knows how to write the web frontend parts, to write the necessary UI to allow users to provision it themselves. Lacking that, it requires someone with access to the bridge admin console, to generate per-user URLs so users can set up the linkage themselves. As far as I know, the only person with such access remains to be me.

**Comment by leonerd** [comment]
Though I suppose without that UI in Riot et.al. there could be a provisioning method involving sending a direct message to the bridge bot user somehow. That would be a new bit of code flow that would need adding to the bridge, but would bypass the dependency on frontend UI developers.

**Comment by joenas** [comment]
There's work being done on other bridges to add a provisioning. See https://github.com/turt2live/matrix-appservice-webhooks/blob/master/PROVISIONING.md for an example! 
I have a bridge myself where I'd like to add it but lack the time atm :/

**Comment by devnoname120** [comment]
> Though I suppose without that UI in Riot et.al. there could be a provisioning method involving sending a direct message to the bridge bot user somehow. That would be a new bit of code flow that would need adding to the bridge, but would bypass the dependency on frontend UI developers.

This would be great. And once it is implemented, the UI could follow.

**Comment by yangm97** [comment]
Bump?

**Comment by JuniorJPDJ** [comment]
bump? ;/

**Comment by ildar** [comment]
Gitter seem to finish their part of the job of replacing matrixbot with something user-name contained: https://gitlab.com/gitlab-org/gitter/webapp/-/issues/1076
When our bridge would start using the feature?
